proposals/mainnet_migration.md

```diff
@@ -0,0 +1,540 @@
+
+#### Later Versions of Lisk Core v4
+
+Once the snapshot block is final, a patched version of Lisk Core v4 will be released in which the block ID of the snapshot block is hard coded. This allows to validate a downloaded snapshot block by simply checking its block ID.
+
+#### Downloading from Server
+
+The URL can be configured. By default, the URL points to a server of the Lisk Foundation. The snapshot block of this hard fork is validated by checking that its block ID matches the hard coded block ID in the patched version of Lisk Core v4\. The snapshot block of the previous migration is validated by checking that its block ID matches with the `previousBlockID` property of its child block (see also [here](#getting-the-lisk-core-3-block-history)). Hence, there are no security concerns by downloading them from a non-trusted server. Users can nevertheless configure their node to download it from a server of their choice. The snapshot blocks could be provided by any users. This may also be helpful for situations where the Lisk Foundation server should be down.
+
+### Assigning Generator Key to Validator Accounts
+
+In the protocol used by Lisk Core v4, each validator needs a [generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) in order to generate blocks. Moreover, the DPoS module expects that each validator account has such a key specified in the snapshot block. Otherwise, the snapshot block is invalid (see [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-state-initialization)). Therefore, it is aimed to use the existing public keys of the validator accounts as the initial generator keys in the snapshot block. However, the public key does not belong to the [account state](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html) anymore with Lisk Core v3\. For this reason, the whole block history from the previous snapshot block onwards is scanned for public keys of registered validators. The found keys are used as generator keys in the snapshot block. Validator accounts for which no public key was found get an "invalid" public key assigned, i.e., one for which every signature validation will fail. The specific value is discussed in the [following subsection](#invalid-ed25519-public-key). Those validators should send an [update generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#update-generator-key) transaction as soon as possible once Lisk Core v4 is used, ideally within the [bootstrap period](#bootstrap-period). Otherwise, those validators can still be active validators, but are not able to generate valid blocks as they cannot create [valid block signatures](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md#signature). In the worst case, this may lead to banning which happens after 50 blocks in a row are missed.
```

Comment:
```suggestion
In the protocol used by Lisk Core v4, each validator needs a [generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) in order to generate blocks. Moreover, the DPoS module expects that each validator account has such a key specified in the snapshot block. Otherwise, the snapshot block is invalid (see [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-state-initialization)). Therefore, it is aimed to use the existing public keys of the validator accounts as the initial generator keys in the snapshot block. However, the public key does not belong to the [account state](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html) anymore with Lisk Core v3\. For this reason, the whole block history from the previous snapshot block onwards is scanned for public keys of registered validators. The found keys are used as generator keys in the snapshot block. Validator accounts for which no public key was found are assigned an "invalid" public key, i.e., one for which every signature validation will fail. The specific value is discussed in the [following subsection](#invalid-ed25519-public-key). These validators should send an [update generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#update-generator-key) transaction as soon as possible once Lisk Core v4 is used, ideally within the [bootstrap period](#bootstrap-period). Otherwise, they can still be active validators, but not able to generate valid blocks as they cannot create [valid block signatures](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md#signature). In the worst case, this may lead to banning which happens after 50 blocks are missed in a row.
```